### ISSUE TITLE
Optimize massive concurrent requests

### ISSUE THREAD

**Issue by Mccranky83** [Mccranky83/byw-mangadl#1]
Initially, I was sending approximately 5,000+ fetch requests in parallel to retrieve manga images. This led to several issues:

1. Browser instability due to excessive concurrent connections
2. Frequent `net::ERR_INSUFFICIENT_RESOURCES` errors

#### First Iteration: Grouped Requests
My initial attempt to mitigate these problems involved grouping requests:

1. Dictate a group of a maximum number of `max_par`
2. Sent groups sequentially, allowing individual requests within a group to run concurrently

```js
async function limitParDl(items, fn, max_par) {
  let lock = false;
  let groups = items.reduce((acc, cur, i) => {
    !(i % max_par) && acc.push([]);
    acc[acc.length - 1].push(cur);
    return acc;
  }, []);
  await Promise.all(
    groups.map(async (cur) => {
      while (lock) {
        await new Promise((res) => {
          setTimeout(res, 1000);
        });
      }
      lock = true;
      await Promise.all(
        cur.map(
          (cur) =>
            new Promise(async (res) => {
              await fn(cur);
              res();
            }),
        ),
      );
      lock = false;
    }),
  );
}
```
While this improved stability, it introduced significant idle time between groups, reducing overall efficiency.

#### Current Solution - Semaphore-based Throttling
Then I came up with a more efficient solution using a semaphore:

```javascript
class Semaphore {
  constructor(max_par) {
    this.counter = max_par;
    this.waitlist = [];
  }
  async acquire() {
    if (this.counter > 0) {
      this.counter--;
      return;
    }
    await new Promise((res) => this.waitlist.push(res));
  }

  release() {
    this.counter++;
    if (this.waitlist.length > 0) {
      this.counter--;
      this.waitlist.shift()();
    }
  }
}

async function limitParDl(items, fn, args, max_par) {
  const s = new Semaphore(max_par);
  await Promise.all(
    items.map(async (cur) => {
      await s.acquire();
      await fn(cur, ...args).finally(s.release.bind(s));
    }),
  );
}
```

This approach maintains a consistent level of concurrency (`max_par`), and minimizes idle time compared to grouped sequential approach.

#### Implementation Details

Due to the immediate execution of fetch requests after definition, I couldn't define all requests upfront, so I implemented a waitlist system. This system pre-fills a queue with resolve methods for dummy promises, each representing a fetch request that exceeds the maximum concurrency limit (`max_par`). As slots become available, the code executes the resolve method, which defines and initiates the next fetch request in line.

So, to sum up:

1. New requests are only initiated when slots become available
2. All requests are processed efficiently without unnecessary delays
